### DISCUSSION TITLE
sort children by date

### DISCUSSION THREAD

**Discussion by Olifair** [freeplane/freeplane#1658] [Ideas]
Hello, 
is there any way to sort children by creation or modification date ? 
Thanks
Oliver

**Comment by i-plasm** [comment]
Hi @Olifair , you can do that via a simple script. The script below sorts immediate children (i.e not grandchildren) by creation date.  With a simple change, you can create a second script for sorting by modification date (using `getLastModifiedAt()` instead of `getCreatedAt()`). You can save to a file called `sortChildrenByCreationDate.groovy`  and install it to Freeplane, then assign a shortcut for quick access.

~~~ groovy
import javax.swing.JOptionPane
import org.freeplane.core.ui.components.UITools
import org.freeplane.plugin.script.proxy.ScriptUtils

def selecteds = ScriptUtils.c().getSelecteds()

if (selecteds.size() > 1) {
  JOptionPane.showMessageDialog(UITools.getCurrentFrame(),  "You have more than one node selected.  Please select only one node, and try again", "Warning", JOptionPane.WARNING_MESSAGE)
  return
}

selecteds[0].sortChildrenBy{it.getCreatedAt()}
~~~

**Reply by Olifair** [reply]
Awsome, thanks very much.  How would sort in reverse order ? 
I had a look at : 
https://docs.freeplane.org/api/org/freeplane/plugin/script/proxy/Proxy.Node.html
but with no indication for an option in this command for the sort order. 

I found : 
sorted.reverse().eachWithIndex
here : 
https://docs.freeplane.org/scripting/Scripts_collection.html#sort-child-nodes-alphabetically-by-length-or-by-other-properties
but this is not the same command as used in the script above.

**Reply by i-plasm** [reply]
Sure! For reverse order we can use a Date wrapper:

~~~ groovy
import javax.swing.JOptionPane
import org.freeplane.core.ui.components.UITools
import org.freeplane.plugin.script.proxy.ScriptUtils

def selecteds = ScriptUtils.c().getSelecteds()

if (selecteds.size() > 1) {
  JOptionPane.showMessageDialog(UITools.getCurrentFrame(),  "You have more than one node selected."  +
      " Please select only one node, and try again", "Warning", JOptionPane.WARNING_MESSAGE)
  return
}

selecteds[0].sortChildrenBy{
  new DescendingDateWrapper(it.getCreatedAt())
}

class DescendingDateWrapper implements Comparable<DescendingDateWrapper> {
  Date wrappedDate

  DescendingDateWrapper(Date date) {
    this.wrappedDate = date
  }

  @Override
  public int compareTo(DescendingDateWrapper anotherDate) {
    return anotherDate.wrappedDate.compareTo(wrappedDate)
  }
}
~~~

**Reply by Olifair** [reply]
Thanks very much, this makes my life (and working with my ToDo map) so much easire. 
It is quite interesting to see that the small functional chnage of sorting in reverse order needs quite a different script. 
cheers
Oliver

**Reply by i-plasm** [reply]
You're welcome. Actually reversing can be really easy, like the script you shared above from the Scripts Collections, where you only have to include `reverse()` in `sorted.reverse().eachWithIndex` - I can not do that with the script I wrote because it works differently. You can use that script if you need any other type of ordering (not just dates), but the `// @ExecutionModes({ON_SELECTED_NODE, ON_SELECTED_NODE_RECURSIVELY})` line will apply it to all descendants. If you want to apply ordering on just the children, you would have to remove that line.

**Reply by Olifair** [reply]
Thanks for the explications. :-)
